implement certificate installation for REST (#1404)
corda/corda-docs-portal: content/en/platform/corda/5.0/deploying-operating/config/tls-certification.md

The REST Worker TLS certificate is presented to a client any time a HTTPS connection is made.
If no specific parameters are provided, the self-signed certificate is used and the connection to the REST Worker is always HTTPS. However, a warning will be emitted into the REST log explaining how to provide parameters for custom TLS certificates.
There are a two ways that a TLS certificate can be made available to the REST Worker. For either method there are three pieces of information required to install a valid TLS certificate.

* The TLS certificate itself is signed by a Certification Authority (CA) or an intermediary.
* A private key corresponding to the public key is included into the TLS certificate.
* The Certification Chain leads up to CA.

## Certificate Information Provided in PEM Format

REST Worker now accepts optional command line arguments: 

* PEM representation of a TLS certificate to be used: `-rtls.crt.path=<path to tls.crt>`

* PEM representation of the private key and of the CA certificate. This could also be a certification path where CA certificate is listed as the last one on the file: `-rtls.key.path=<path to tls.key>`

## Use with Helm Chart

> When deploying the C5 cluster to Kubernetes using Helm, certificate information can be provided as a Kubernetes secret. 

Reviewer: ```suggestion
When deploying a Corda cluster to Kubernetes using Helm, certificate information can be provided as a Kubernetes secret. 
```

One of the reasons I thought this might belong in the Deployment section. Seems like something that needs to be considered/performed as part of the deployment process

> The Cluster administrator can either create a Kubernetes secret manually to hold the certificate information or allow Helm to generate a new secret.

Reviewer: ```suggestion
You can either create a Kubernetes secret manually to hold the certificate information or allow Helm to generate a new secret.
```

(My bad suggestion, I think, apologies!)

> 

Reviewer: ```suggestion
```

You can specify the secret name in the Helm values file under the parameter  `Values.workers.rest.tls.secretName`. If this optional value is not provided, Helm generates the certificate data at installation time and automatically creates a Kubernetes secret for the REST Worker to use.

{{< note >}}
If the TLS certificate has been installed as a Kubernetes secret and subsequently secret data modified, REST Worker Pod will not currently detect a change in the TLS certificate data until Pod is restarted.
{{</ note >}}

## Certificate Information Provided as PKCS12 Key Store

Alternatively, it is possible to supply information as a password protected PKCS12 key store.

> To pass a PKCS12 key store to the REST Worker, the following start-up arguments can be used:

Reviewer: See previous comments for line 25

---
title: "TLS Certification Installation on REST Worker"
version: 'Corda 5.0'
date: '2023-05-29'
menu:
  corda5:
    identifier: corda5-tls-certification-installation
    parent: corda5-cluster-config
    weight: 3075

> section_menu: corda5

Reviewer 2: Can you validate that this front matter is correct ?
We have both:
```
menu:
  corda5:
```
and also:
```
section_menu: corda5
```

Author: The front matter isn't fully understood but the front matter variables are used in other files and work correctly!

---

# TLS Certification Installation on REST Worker

> The Corda 5 GA release the TLS certificate on REST Worker is presented to a client any time HTTPS connection is made.

Reviewer: ```suggestion
The REST Worker TLS certificate is presented to a client any time a HTTPS connection is made.
```
(I think this is what is being said.. )

> Before, a self-signed certificate shipped with the REST Worker. If no specific parameters are provided, the self-signed certificate will be used as a default option such that the connection to the REST Worker is always HTTPS. However, a warning will be emitted into the REST log explaining how to provide parameters for custom TLS certificates.

Reviewer: I don't think this first sentence is necessary?

> Before, a self-signed certificate shipped with the REST Worker. If no specific parameters are provided, the self-signed certificate will be used as a default option such that the connection to the REST Worker is always HTTPS. However, a warning will be emitted into the REST log explaining how to provide parameters for custom TLS certificates.

Reviewer: ```suggestion
If no specific parameters are provided, the self-signed certificate is used and the connection to the REST Worker is always HTTPS. However, a warning will be emitted into the REST log explaining how to provide parameters for custom TLS certificates.
```

> There are a few ways that a TLS certificate can be made available to the REST Worker. For either method there are three pieces of information required to install a valid TLS certificate.

Reviewer: "either" are there two or a few?

> There are a few ways that a TLS certificate can be made available to the REST Worker. For either method there are three pieces of information required to install a valid TLS certificate.

Reviewer: I would list the "ways" as bulleted cross-refs to the sections.

> 

Reviewer: Should this be an unordered list?

> 1. The TLS certificate itself is signed by a Certification Authority (CA) or an intermediary;

Reviewer: ```suggestion
1. The TLS certificate itself is signed by a Certification Authority (CA) or an intermediary.
```

2. A private key corresponding to the public key is included into the TLS certificate;
3. The Certification Chain leads up to CA

## Certificate Information Provided in PEM Format

> REST Worker now accepts optional command line arguments: 

Author: I asked Victor about this and he said: These should be fairly familiar terms to any enterprise software operator/DevOps engineer.
As the documentation says:
"TLS certificate"
"private key"
I am not sure how much more clear we can make this. The only idea comes to mind is to include links to Wikipedia articles..."  I can add the links if you think it will make it clearer?

1. PEM representation of a TLS certificate to be used: `-rtls.crt.path=<path to tls.crt>`

2. PEM representation of the private key and of the CA certificate. This could also be a certification path where CA certificate is listed as the last one on the file: `-rtls.key.path=<path to tls.key>`

## Use with Helm Chart

> When deploying the C5 cluster to Kubernetes (K8s) using Helm, certificate information can be provided as a K8s secret. 

Reviewer: ```suggestion
When deploying the C5 cluster to Kubernetes using Helm, certificate information can be provided as a Kubernetes secret. 
```

The C5 Cluster operator can either create K8s secret which hold certificate information themselves or allow Helm to generate a new secret.

Our Helm values file now has a parameter to specifysecret name under `Values.workers.rest.tls.secretName`. If this optional value is not provided, Helm will generate certificate data at installation time and will automatically create K8s secret for REST Worker to use.

{{< note >}}
If the TLS certificate has been installed as a K8s secret and subsequently secret data modified, REST Worker Pod will not currently detect a change in the TLS certificate data until Pod is restarted.
{{</ note >}}

## Certificate information provided as PKCS12 key store

> Alternatively, it is possible to supply such information as password protected PKCS12 key store.

Reviewer: Replace "such information"

> In order to pass PKCS12 key store to REST Worker, the following start-up arguments can be used:

Reviewer: ```suggestion
To pass a PKCS12 key store to the REST Worker, the following start-up arguments can be used:
```
Probably just my lack of knowledge but it's not clear to me where they would be set. Might be worth clarifying.

> 

Reviewer: Unordered list?

1. The full path to a key store file: `-rtls.keystore.path=<path to a keystore>`

2. Password is used for opening the key store: `-rtls.keystore.password=<password>`

{{< note >}}

> It is best practice to pass the password through an environment variable.

Reviewer: ```suggestion
R3 recommend passing the password through an environment variable.
```
